src/bosh_azure_cpi/lib/cloud/azure/vm_manager.rb

```diff
@@ -4,33 +4,64 @@ module Bosh::AzureCloud
+    def initialize(azure_config, registry_endpoint, disk_manager, disk_manager2, azure_client2, storage_account_manager, stemcell_manager, stemcell_manager2, light_stemcell_manager)
       @azure_config = azure_config
       @registry_endpoint = registry_endpoint
       @disk_manager = disk_manager
       @disk_manager2 = disk_manager2
       @azure_client2 = azure_client2
       @storage_account_manager = storage_account_manager
-      @use_managed_disks = @azure_config['use_managed_disks']
+      @stemcell_manager = stemcell_manager
+      @stemcell_manager2 = stemcell_manager2
+      @light_stemcell_manager = light_stemcell_manager
+      @use_managed_disks = azure_config['use_managed_disks']
       @logger = Bosh::Clouds::Config.logger
     end
 
-    def create(instance_id, location, stemcell_info, vm_properties, network_configurator, env)
+    def create(instance_id, location, stemcell_id, vm_properties, network_configurator, env)
       # network_configurator contains service principal in azure_config so we must not log it.
-      @logger.info("create(#{instance_id}, #{location}, #{stemcell_info.inspect}, #{vm_properties}, ..., ...)")
+      @logger.info("create(#{instance_id}, #{location}, #{stemcell_id}, #{vm_properties}, ..., ...)")
       resource_group_name = instance_id.resource_group_name()
       vm_name = instance_id.vm_name()
-      vm_size = vm_properties.fetch('instance_type', nil)
 
       # When both availability_zone and availability_set are specified, raise an error
       cloud_error("Only one of 'availability_zone' and 'availability_set' is allowed to be configured for the VM but you have configured both.") if !vm_properties['availability_zone'].nil? && !vm_properties['availability_set'].nil?
+      zone = vm_properties.fetch('availability_zone', nil)
+      unless zone.nil?
+        cloud_error("'#{zone}' is not a valid zone. Available zones are: #{AVAILABILITY_ZONES}") unless AVAILABILITY_ZONES.include?(zone.to_s)
+      end
 
       check_resource_group(resource_group_name, location)
 
+      stemcell_info = get_stemcell_info(stemcell_id, vm_properties, location)
+
+      # When availability_zone is specified, VM won't be in any availability set;
+      # Otherwise, VM can be in an availability set specified by availability_set or env['bosh']['group']
+      availability_set_name = vm_properties['availability_zone'].nil? ? get_availability_set_name(vm_properties, env) : nil
+
+      primary_nic_tags = AZURE_TAGS.dup
+      # Store the availability set name in the tags of the NIC
+      primary_nic_tags['availability_set'] = availability_set_name unless availability_set_name.nil?
+      network_interfaces = create_network_interfaces(resource_group_name, vm_name, location, vm_properties, network_configurator, primary_nic_tags)
+
+      availability_set = get_or_create_availability_set(resource_group_name, availability_set_name, vm_properties, location)
```

Comment:
should this line be moved to line 41?

Reply:
NICs also use the name of availability set as a tag